### PR TITLE
make nn.Module generic at compile time

### LINKED ISSUE #140458: Making nn.Module generic
### 🐛 Describe the bug

I was chatting with the type checking team at Meta and I think we have line of sight to making nn.Module generic.

To recap the constraints:

* We cannot make nn.Module inherit from Generic because this changes the metaclass and is breaking
* We prefer to NOT have a pyi file for module.py
* We need to support downstream use of nn.Module in type signatures without future annotations

Here is the recipe:

* We use a `if TYPE_CHECKING` block to swap between the Generic and non-Generic version of nn.Module. The Generic version can inherit from Generic but it will have no runtime effect because it is TYPE_CHECKING only. It is somewhat irritating we have to reproduce the types for all methods on nn.Module but this seems unavoidable. We have to pay for an indentation tax for this
* We add a `__class_getitem__` on Module which does nothing so that direct use of `nn.Module[Blah]` works
* We use ParamSpec https://peps.python.org/pep-0612/ to allow for accurately typing arguments/return type

I will happily shepherd a PR if someone wants to make an attempt at this.

### Versions

main

### PR DESCRIPTION
**Pull request by aniltolwani**
Fixes #140458.

# Make nn.Module generic without breaking changes

This PR adds generic type support to `nn.Module` in a backward-compatible way, enabling better static type checking for PyTorch models.

## Motivation

Currently, `nn.Module` lacks generic type parameters, making it impossible to properly type-check module inputs and outputs at the class level. This leads to:
- Type errors only being caught at runtime
- Difficulty in expressing and verifying type relationships between connected modules
- Less clear module interfaces

## Implementation

The change uses the `TYPE_CHECKING` pattern (also used by FastAPI, Pydantic, and other major Python libraries) to provide generic typing without runtime impact:

```python
P = ParamSpec('P')
T_co = TypeVar('T_co', covariant=True)

if TYPE_CHECKING:
    class Module(Generic[P, T_co]):
        def forward(self, *args: P.args, **kwargs: P.kwargs) -> T_co: ...
else:
    class Module:
        @classmethod
        def __class_getitem__(cls, item):
            return cls
```

This enables type-safe module definitions, i.e.:
```python
class MyModule(nn.Module[Tuple[Tensor], Tensor]):
    def forward(self, x: Tensor) -> Tensor:
        return x * 2

# Type checker catches errors:
class BadModule(nn.Module[Tuple[str], int]):  # Error: wrong types
    def forward(self, x: str) -> int:
        return 42
```

cc @ezyang @malfet @xuzhao9 @gramster

### CODE REVIEW

**Comment by aniltolwani** [comment]
@pytorchbot label "release notes: nn"

This PR adds generic type support to `nn.Module`, enabling better static type checking. This is a user-facing change that affects the [nn](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py) module's typing capabilities.

**Comment by ezyang** [comment]
Alas, you will need to indent the entire class, not just what you have done her

**Comment by aniltolwani** [comment]
@ezyang ah thanks, I totally missed it. I changed it now, unfortunately this makes the diff messy but the only real change is indenting the module defn.

**Comment by ezyang** [comment]
Some other tests need updating it looks like

**Comment by bluenote10** [comment]
As far as I can see, in the current state this PR would be a big step backwards in terms of type information on `torch.nn.Module`, because the majority of the functions are missing in the `TYPE_CHECKING` version. Won't this break type checking on user side in a pretty dramatic way like that?

Even if duplicating all existing signatures into the `TYPE_CHECKING` version, I'm not sure if this is a good idea. I'm wondering if we then should purposefully remove all type annotation from the non `TYPE_CHECKING` version to avoid having "dead/fake annotations" there. After all, all these annotations are _not used_ and could be entirely wrong and misleading.

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -399,2595 +404,2602 @@ def _forward_unimplemented(self, *input: Any) -> None:
...
-        """
-        return self._apply(lambda t: t.type(dst_type))
+            Returns:
+                Module: self
+            """
+            return self._apply(lambda t: t.xpu(device))
 
-    def float(self: T) -> T:
-        r"""Casts all floating point parameters and buffers to ``float`` datatype.
+        def mtia(self: T, device: Optional[Union[int, device]] = None) -> T:
```
Comment (bluenote10) [inline torch/nn/modules/module.py:1105]: Perhaps to clarify why I think this doesn't really make sense: As far as I can see, these type annotations here are all pointless, because they do not appear inside the `if TYPE_CHECKER` branch. No type checker will ever care about them. We could replace, `device: Optional[Union[int, device]]` by `device: SomeGarbageAnnotation` and we will never notice. Even worse: Inside the function we could do `device.call_nonexisting_function()` and we would also never notice, because the type checker doesn't look at anything here.

And super bad for users: If we forget to move these signatures into the `if TYPE_CHECKING` block properly, it will become possible to release a pytorch version that has an incomplete/broken interface in the type stubs. After all, it might be possible to get the pytorch CI "green" by only typing a subset of the methods that are used within one of the type-checked usages inside pytorch. But if some method is never actively used, it can be omitted in the stubs, and users would get an undefined method/attribute error when trying to use an actually existing method.

Comment (YodaEmbedding) [reply torch/nn/modules/module.py:1105]: Having an extra indent by putting it inside an `if`/`else` is... fishy.

Perhaps better is to (somehow) tell the `*.pyi` stub generator what the desired hint is.

In this case, the PR appears to propose:
```python
# torch/nn/modules/module.pyi

class Module(Generic[P, T_co]):
    def forward(self, *args: P.args, **kwargs: P.kwargs) -> T_co: ...

    # Instead of:
    # forward: Callable[..., Any] = _forward_unimplemented
```
